src/goto-programs/loop_unroll.cpp

```diff
@@ -0,0 +1,213 @@
+
+  t_skip->make_skip();
+  t_skip->location = loop_head->location;
+  t_skip->function = loop_head->function;
+  */
+
+  // If there is an inner control flow we need a map for it
+  std::map<goto_programt::targett, unsigned> target_map;
+  {
+    unsigned count = 0;
+    goto_programt::targett t = loop.get_original_loop_head();
+    t++; // get first instruction of the loop
+    for(; t != loop_exit; t++, count++)
+    {
+      assert(t != goto_program.instructions.end());
+      target_map[t] = count;
+    }
+  }
+
+  // re-direct any branches that go to loop_head to loop_iter
+  // TODO: not sure how to deal with this
+  /*
+  for(goto_programt::targett t = loop_head; t != loop_iter; t++)
+  {
+    assert(t != goto_program.instructions.end());
+    for(auto &target : t->targets)
+      if(target == loop_head)
+        target = loop_iter;
+  }
+  */
+
+  // we make k-1 copies, to be inserted before loop_exit
+  goto_programt copies;
+  for(int i = 1; i < bound; i++)
+  {
+    // make a copy
+    std::vector<goto_programt::targett> target_vector;
+    target_vector.reserve(target_map.size());
+    // IF !COND GOTO X
+    goto_programt::targett t = loop.get_original_loop_head();
+    t++; // get first instruction of the loop
+    for(; t != loop_exit; t++)
+    {
+      assert(t != goto_program.instructions.end());
+      goto_programt::targett copied_t = copies.add_instruction(*t);
+      target_vector.push_back(copied_t);
+    }
+
+    for(unsigned i = 0; i < target_vector.size(); i++)
+    {
+      goto_programt::targett t = target_vector[i];
+      for(auto &target : t->targets)
+      {
+        std::map<goto_programt::targett, unsigned>::const_iterator m_it =
+          target_map.find(target);
+        if(m_it != target_map.end()) // intra-loop?
+        {
+          assert(m_it->second < target_vector.size());
+          target = target_vector[m_it->second];
+        }
+      }
+    }
+  }
+  // now insert copies before loop_exit
+  goto_program.destructive_insert(loop_exit, copies);
+  // TODO: Is this needed? remove_skip(goto_function.body);
+  return true;
+}
+
+int bounded_loop_unroller::get_loop_bounds(loopst &loop)
+{
+  /**
+   * This looks for the following template
+   * 
+   * z: SYMBOL = K0
+   * 
+   * a: IF !(SYMBOL < K)
+   * b: ... // code that only reads SYMBOL
+   * b: SYMBOL++
+   * 
+   * If this is matched properly then set
+   * bound as K - K0 and return true
+   */
+  goto_programt::targett t = loop.get_original_loop_head();
+  symbol2tc SYMBOL;
+  int K = 0;
+
+  // 1. Check the condition. 't' should be IF !(SYMBOL < K) THEN GOTO x
+  if(!t->is_goto())
+    return -1;
+  // Pattern match SYMBOL and K, if the relation is <= then K = K + 1
+  {
+    auto &cond = to_not2t(t->guard).value;
+    if(!is_lessthan2t(cond) && !is_lessthanequal2t(cond))
+      return -1;
+
+    std::shared_ptr<relation_data> relation;
+    relation = std::dynamic_pointer_cast<relation_data>(cond);
+    if(relation)
```

Comment:
Can you explain the purpose of this relation check?

Reply:
@rafaelsamenezes can remove this relation object and call `is_symbol2t` (and the other `is_*`) using the expr's operands by using `cond->get_sub_expr(0)`, `cond->get_sub_expr(1)`, etc.

Reply:
I am trying to extract the condition for the loop which should be a relation